state/caasmodel_test.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state_test
+
+import (
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/names.v2"
+
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/feature"
+	"github.com/juju/juju/state"
+	"github.com/juju/juju/storage"
+	"github.com/juju/juju/testing/factory"
+)
+
+type CAASModelSuite struct {
+	ConnSuite
+}
+
+var _ = gc.Suite(&CAASModelSuite{})
+
+// createTestModelConfig returns a new model config and its UUID for testing.
+func (s *CAASModelSuite) createTestModelConfig(c *gc.C) (*config.Config, string) {
+	return createTestModelConfig(c, s.modelTag.Id())
+}
+
+func (s *CAASModelSuite) newCAASModel(c *gc.C) (*state.Model, names.ModelTag, names.UserTag) {
+	cfg, uuid := s.createTestModelConfig(c)
+	modelTag := names.NewModelTag(uuid)
+	owner := names.NewUserTag("test@remote")
+	model, st, err := s.State.NewModel(state.ModelArgs{
+		Type:      state.ModelTypeCAAS,
+		CloudName: "dummy",
+		Config:    cfg,
+		Owner:     owner,
+	})
+	c.Assert(err, jc.ErrorIsNil)
+	s.AddCleanup(func(*gc.C) { st.Close() })
+	return model, modelTag, owner
+}
+
+func (s *CAASModelSuite) TestNewModel(c *gc.C) {
+	s.SetFeatureFlags(feature.CAAS)
+
+	model, modelTag, owner := s.newCAASModel(c)
+	c.Assert(model.Type(), gc.Equals, state.ModelTypeCAAS)
+	c.Assert(model.UUID(), gc.Equals, modelTag.Id())
+	c.Assert(model.Tag(), gc.Equals, modelTag)
+	c.Assert(model.ControllerTag(), gc.Equals, s.State.ControllerTag())
+	c.Assert(model.Owner(), gc.Equals, owner)
+	c.Assert(model.Name(), gc.Equals, "testing")
+	c.Assert(model.Life(), gc.Equals, state.Alive)
+	c.Assert(model.CloudRegion(), gc.Equals, "")
```

Comment:
Seems wrong that we're exposing CloudRegion for CAAS models if it never applies. Can we move that to IAASModel?

Reply:
CAAS models might get regions in the future AFAIUI

Reply:
then we can expose it when it actually means something

Reply:
Many, many places use just take a model entity/interface and call model.Region()
We'd have to check for model type all over.
Federated kubernetes supports regions and it's only a matter of time till that will be woven into our caas cloud model

Reply:
OK, then. When dealing with multi-region federated Kubernetes, do you specify a region? Or do the clients just auto connect to the closest API server? Couldn't tell if regions in this case are really something that we should be exposing to users, like we do with IaaS models.